spec/models/bottleneck_event_spec.rb

```diff
@@ -140,4 +140,178 @@
+        )
+      ])
+    end
+
+    it "generates MemoryUsage projections" do
+      # Generate 30 days of metrics with a gently rising slope
+      30.times do |i|
+        FactoryBot.create(:metric_rollup_host_daily,
+          :resource     => resource,
+          :timestamp    => i.days.ago.beginning_of_day,
+          :time_profile => time_profile,
+          :min_max      => {
+            :max_derived_memory_used      => 300_000 - i * 10_000,
+            :max_derived_memory_available => 500_000
+          }
+        )
+      end
+
+      described_class.generate_future_events(resource)
+
+      events = described_class.order(:id).to_a
+      expect(events).to match_array([
+        have_attributes(
+          :timestamp  => be_within(0.01).of(20.days.from_now.beginning_of_day),
+          :resource   => resource,
+          :event_type => "MemoryUsage",
+          :severity   => 1,
+          :future     => true,
+          :message    => "Memory - Peak Aggregate Used for Child VMs for Collected Intervals (MB) is projected to reach 488.3 GB (100% of Memory Max Total)"
+        ),
+        have_attributes(
+          :timestamp  => be_within(0.01).of(15.days.from_now.beginning_of_day),
+          :resource   => resource,
+          :event_type => "MemoryUsage",
+          :severity   => 1,
+          :future     => true,
+          :message    => "Memory - Peak Aggregate Used for Child VMs for Collected Intervals (MB) is projected to reach 439.5 GB (90% of Memory Max Total)"
+        ),
+        have_attributes(
+          :timestamp  => be_within(0.01).of(7.days.from_now.middle_of_day),
+          :resource   => resource,
+          :event_type => "MemoryUsage",
+          :severity   => 2,
+          :future     => true,
+          :message    => "Memory - Peak Aggregate Used for Child VMs for Collected Intervals (MB) is projected to reach 366.2 GB (75% of Memory Max Total)"
+        ),
+        have_attributes(
+          :timestamp  => be_within(0.01).of(5.days.ago.beginning_of_day),
+          :resource   => resource,
+          :event_type => "MemoryUsage",
+          :severity   => 3,
+          :future     => true,
+          :message    => "Memory - Peak Aggregate Used for Child VMs for Collected Intervals (MB) is projected to reach 244.1 GB (50% of Memory Max Total)"
+        )
+      ])
+    end
+
+    it "generates DiskUsage projections" do
+      # Generate 30 days of metrics with a gently rising slope
+      30.times do |i|
+        FactoryBot.create(:metric_rollup_host_daily,
+          :resource     => resource,
+          :timestamp    => i.days.ago.beginning_of_day,
+          :time_profile => time_profile,
+          :derived_storage_total => 500_000_000_000,
+          :derived_storage_free => 500_000_000_000 - (300_000_000_000 - i * 10_000_000_000)
+        )
+        FactoryBot.create(:metric_rollup_host_hr,
```

Comment:
This is super-weird.  For some reason max_v_derived_storage_used is defined to look at hourly records, but only the hourly records with the timestamp at the beginning of the day, so I had to _also_ include hourly records here.  Makes no sense to me.